hir/src/builder.rs

```diff
@@ -703,6 +799,121 @@ pub trait InstBuilder<'f>: InstBuilderBase<'f> {
         into_first_result!(self.Unary(Opcode::IntToPtr, ty, arg, span))
     }
 
+    /// This is an intrinsic which derives a new pointer from an existing pointer to an aggregate.
+    ///
+    /// In short, this represents the common need to calculate a new pointer from an existing pointer,
+    /// but without losing provenance of the original pointer. It is specifically intended for use
+    /// in obtaining a pointer to an element/field of an array/struct, of the correct type, given a
+    /// well typed pointer to the aggregate.
+    ///
+    /// This function will panic if the pointer is not to an aggregate type
+    ///
+    /// The new pointer is derived by statically navigating the structure of the pointee type, using
+    /// `offsets` to guide the traversal. Initially, the first offset is relative to the original
+    /// pointer, where `0` refers to the base/first field of the object. The second offset is then
+    /// relative to the base of the object selected by the first offset, and so on. Offsets must remain
+    /// in bounds, any attempt to index outside a type's boundaries will result in a panic.
+    fn getelementptr(mut self, ptr: Value, mut indices: &[usize], span: SourceSpan) -> Value {
+        let mut ty = require_pointee!(self, ptr);
+        assert!(
+            !indices.is_empty(),
+            "getelementptr requires at least one index"
+        );
+
+        // Calculate the offset in bytes from `ptr` to get the element pointer
+        let mut offset = 0;
+        while let Some((index, rest)) = indices.split_first().map(|(i, rs)| (*i, rs)) {
+            indices = rest;
+            match ty {
+                Type::Array(ref element_ty, len) => {
+                    assert!(
+                        index < *len,
+                        "invalid getelementptr: index of {} is out of bounds for {}",
+                        index,
+                        ty
+                    );
+                    let element_size = element_ty.size_in_bytes();
+                    let min_align = element_ty.min_alignment();
+                    let align_offset = element_size % min_align;
+                    let padded_element_size = if align_offset != 0 {
+                        element_size + (min_align - align_offset)
+                    } else {
+                        element_size
+                    };
```

Comment:
It's slightly different, since what we're doing here is calculating the size up to the element being selected. I do think there would be some benefit to having some helpers for some of these alignment calculations though, since they occur in a few places.